### PR TITLE
DRAFT: Async summarization of DDSs

### LINKED ISSUE #7934: Async Summarization (November)
## Work Item
Completion of #7615 <- see here for full details

**Approach**
The required work for this item is to break the graph visitation into two passes. The first pass visits the IChannels, instructing the channel to capture whatever state it will need in order to summarize itself. Because attachment occurs on the user's client (vs. the designated summarization client), this pass must remain synchronous to avoid local mutations while the snapshot is being captured.

The second pass revisits the same graph of IChannels, requesting that they produce an ISummaryTree from their previously captured state. This second pass is asynchronous, which allows allows the DDS implementer to upload or download blobs, and/or periodically yield.

What the DDS captures during the first pass is up to the DDS. DDSes using functional/persistent/immutable data structures may perform a cheap clone during this phase, while other DDSes may continue to synchronously produce an ITree in the first pass and simply return it in the second pass.

### PR DESCRIPTION
**Pull request by scarlettjlee**
See issue #7615

Split summarization of DDSs into two parts:
1. synchronous method to capture state needed to summarize the DDS at this point in time
2. asynchronous method to turn the captured state into a summary

Questions for reviewers:
1. Should the breaking change in common be staged over several versions? I'm not sure of the best way to do this when a method becomes async.
2. Should GC also have two methods instead of a single one [to prevent accidentally making something async that should be synchronous]?
3. Should I leave snapshotCore for a version or two before deprecating? or remove it now?
4. Is there anything I should make sure I test?

Not all the existing DDSs and test code are fixed up yet and this change will need more testing.

### CODE REVIEW

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 1]
```diff
@@ -192,39 +192,53 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+        }
+        // eslint-disable-next-line @typescript-eslint/no-unsafe-return
+        return state;
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).summarizeState}
+     */
+    public async summarizeState(capture): Promise<ISummaryTreeWithStats> {
+        assert(this._isSummarizing, 0x077 /* "Possible re-entrancy! Summary should have been in progress." */);
```
Comment (scarlettjlee) [inline packages/dds/shared-object-base/src/sharedObject.ts:220]: Is this okay that `captureSummaryState` sets `_isSummarizing=true` and `summarizeState` sets `_isSummarizing=false`? Is there some legitimate case where the calls are not always paired or someone could get `serializer` in between?

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 2]
```diff
@@ -192,42 +192,35 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
     /**
-     * {@inheritDoc (ISharedObject:interface).summarize}
+     * {@inheritDoc (ISharedObject:interface).captureSummaryState}
      */
-    public summarize(fullTree: boolean = false, trackState: boolean = false): ISummaryTreeWithStats {
-        // Set _isSummarizing to true. This flag is used to ensure that we only use SummarySerializer (created below)
-        // to serialize handles in this object's data. The routes of these serialized handles are outbound routes
-        // to other Fluid objects.
-        assert(!this._isSummarizing, 0x076 /* "Possible re-entrancy! Summary should not already be in progress." */);
-        this._isSummarizing = true;
```
Comment (scarlettjlee) [inline packages/dds/shared-object-base/src/sharedObject.ts:202]: I didn't see any need to prevent the use of `summarizer` during regular summarization and changed it to only check for if we're garbage collecting.

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 3]
```diff
@@ -238,12 +231,13 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
-    protected getGCDataCore(): IGarbageCollectionData {
+    protected async getGCDataCore(): Promise<IGarbageCollectionData> {
         // We run the full summarize logic to get the list of outbound routes from this object. This is a little
-        // expensive but its okay for now. It will be updated to not use full summarize and make it more efficient.
+        // expensive but it's okay for now. It will be updated to not use full summarize and make it more efficient.
         // See: https://github.com/microsoft/FluidFramework/issues/4547
         const serializer = new SummarySerializer(this.runtime.channelsRoutingContext);
-        this.snapshotCore(serializer);
+        const capture = this.captureSummaryStateCore(serializer, false);
+        await this.summarizeStateCore(serializer, capture);
```
Comment (scarlettjlee) [inline packages/dds/shared-object-base/src/sharedObject.ts:240]: Is it okay to prevent use of `serializer` during this async call?

File: packages/dds/cell/src/cell.ts [hunk 4]
```diff
@@ -229,6 +231,25 @@ export class SharedCell<T = any> extends SharedObject<ISharedCellEvents<T>>
         return tree;
     }
 
+    /**
+     * {@inheritDoc @fluidframework/shared-object-base#SharedObject.captureSummaryStateCore}
+     */
+     protected captureSummaryStateCore(serializer: IFluidSerializer, fullTree: boolean): any {
```
Comment (vladsud) [inline packages/dds/cell/src/cell.ts:237]: it would be great not to use 'any'

Comment (vladsud) [reply packages/dds/cell/src/cell.ts:237]: I'd also prefer naming to reflect future possible usage / scenario expansion. I.e. this function captured revision, or state of DDS at given point in time. What can be done with revision / snapshot - will change over time, I do not think producing summary will be the only thing we can do with it. So, I'd rather change name not to have anything to do with summary here. 

This forces an interesting question - should fullTree be an argument to this function, or the method on returned object that produces summary. I think it should be the later, but worth exploring what it means in praxtice.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 2 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:202]: Can you please glance at history of when this was added / what it was supposed to protect against? And is it still valid to assert it even for smaller scope (GC)? I think it does, but would be nice to update comments to clearly articulate why reentrancy is bad / what will break if reentranacy were to happen

File: packages/dds/cell/src/cell.ts (hunk 4 above)
Comment (vladsud) [reply packages/dds/cell/src/cell.ts:237]: I have same feeling for serializer arg. The intention of this function is to be fast and do minimal processing for something like property DDS that is based on immutable data. So I think anything that has relationship to summary / long processing should be in second method / method of object that this call should return.

This changes flow substantially. So I'm not sure about all the side-effects of this direction, but I'd try to see where it leasds.

Note that we can always split SharedObject into two layer - one base that has no assumptions and supports all flows.
And then derived abstract class (to be used by all DDSs other than PropertyDDS) that would hide these complexities (in the same way as SharedObject does today) and just expose single synchronous method to summarize exactly how it works today, creating serializer in first call (and keeping it around), and exposing summary and gc info at a later stage.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 3 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:240]: I'm not sure what you mean. But looking at this function, it feels like it does not care about reentrancy at all (same for getGCData), so maybe we remove _isGCing checks?

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 2 above)
Comment (DLehenbauer) [reply packages/dds/shared-object-base/src/sharedObject.ts:202]: Here you go: https://github.com/microsoft/FluidFramework/pull/3975/files#diff-97e77be555c966d15b3e7d746dbbc24daf3e5950606ff395cc88638119f40620R93

File: packages/dds/cell/src/cell.ts (hunk 4 above)
Comment (scarlettjlee) [reply packages/dds/cell/src/cell.ts:237]: I agree with getting rid of `serializer` here. It does mean bigger changes to all the DDS implementations though, since I can't just call existing snapshotCore implementations.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 2 above)
Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObject.ts:202]: Based on the history, I see a series of changes where getting GC data during summarization was more entangled and then some of it removed without also removing the protection of _isSummarizing and usage of SummarySerializer. My interpretation of the current code is that during GC (which calls summarization code), we still want to prevent the use of any other serializer except the SummarySerializer passed in. But during real summarization, we don't care about this at all. And we don't otherwise care about re-entrancy except inasmuch as it messes with the value of _isSummarizing/GCing. @agarwal-navin, is this correct?

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 3 above)
Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObject.ts:240]: I meant the serializer getter, which we don't want used during GC. My concern is that making this async might allow some other code [not related to GC process] to run in the meantime which legitimately calls the serializer getter.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 2 above)
Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:202]: Yeah, that sounds about right. We need to use the `SummarySerializer` during GC because we want to get a list of all handles serialized during that process. Since summarize is done after getting the GC data, it's fine to remove this during summarize.

File: common/lib/container-definitions/src/runtime.ts [hunk 5]
```diff
@@ -97,7 +97,7 @@ export interface IRuntime extends IDisposable {
      * using IDs generated locally. After attach, these IDs cannot be used, so this table maps the old local IDs to the
      * new storage IDs so requests can be redirected.
      */
-    createSummary(blobRedirectTable?: Map<string, string>): ISummaryTree;
```
Comment (vladsud) [inline common/lib/container-definitions/src/runtime.ts:100]: I want to make sure it works correctly, and comments to reflect expectations.
This API should return state of container at the time when this call was initiated, even though promise may take a minute to get completed.
Given that it's super easy to create bugs here with async workflow (because user edits in between start of this call and when promise is completed can get into snapshot and thus make snapshot incorrect), so I'd expect to see extensive tests validating various types of changes (DDS, but also structural, i.e., creation of DDS / data stores) in between those two points never get into snapshot.

File: packages/dds/cell/src/cell.ts (hunk 4 above)
Comment (vladsud) [reply packages/dds/cell/src/cell.ts:237]: You can avoid a lot of changes - I believe I provided feedback earlier in some of the comments, but having two base layers here - one that makes all these changes and allows full flexibility, and then abstract implementation of that protocol that requires existing synchronous summarization process for all DDSs other than those that want more control.
Happy to chat about it, but I think creating more structure here (i.e. returning an object) is a must

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 2 above)
Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObject.ts:202]: I update the comments related to _isGCing to clarify what it's for and why re-entrancy is a problem.

File: packages/dds/cell/src/cell.ts (hunk 4 above)
Comment (scarlettjlee) [reply packages/dds/cell/src/cell.ts:237]: Something like this commit? [https://github.com/microsoft/FluidFramework/pull/8422/commits/31462ac7f56fced0e9bf47642143dc9702951cce](url)

The diff is not very helpful. Basically, SharedObjectBase has:
```typescript
interface ISharedObjectSnaphost {
    summarize(serializer: IFluidSerializer, fullTree?: boolean): Promise<ISummaryTreeWithStats>;
}
class SharedObjectBase {
...
    public captureState(): ISnaphost {
        const state = this.captureStateCore();
        return { summarize: async (fullTree: boolean = false) => {
            return state.summarize(this._serializer, fullTree);
        }};
    }
    protected abstract captureStateCore(): ISharedObjectSnaphost;
}
```
and SharedObject only has:
```typescript
class SharedObject extends SharedObjectBase {
    protected captureStateCore(): ISharedObjectSnaphost {
        const snapshot = this.snapshotCore(this.serializer);
        return { summarize: async (serializer: IFluidSerializer, fullTree: boolean = false) => {
            return convertToSummaryTreeWithStats(snapshot, fullTree);
        }};
    }
    protected abstract snapshotCore(serializer: IFluidSerializer): ITree;
}
```
With fullTree moved to the async method, it wouldn't be possible to have [renamed] snapshotCore return ISummaryTreeWithStats (#8013).

Comment (vladsud) [reply packages/dds/cell/src/cell.ts:237]: Yep, something like that

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 6]
```diff
@@ -3,450 +3,46 @@
...
+import { ITree } from "@fluidframework/protocol-definitions";
+import { convertToSummaryTreeWithStats } from "@fluidframework/runtime-utils";
+import { ISharedObjectStateHost, SharedObjectBase } from "./sharedObjectBase";
+import { ISharedObjectEvents } from "./types";
 
 /**
- *  Base class from which all shared objects derive
+ *  Base class for shared objects with synchronous summarization
  */
 export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedObjectEvents>
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:19]: It would be easier to see the diff if both SharedObject & SharedObjectBase would stay in this file initially, and future PR would be mechanical movement of code to another file

File: packages/dds/shared-object-base/src/sharedObjectBase.ts [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    IGarbageCollectionData,
+    ISummaryTreeWithStats,
+} from "@fluidframework/runtime-definitions";
+import { FluidSerializer } from "@fluidframework/runtime-utils";
+import { ChildLogger, EventEmitterWithErrorHandling } from "@fluidframework/telemetry-utils";
+import { SharedObjectHandle } from "./handle";
+import { SummarySerializer } from "./summarySerializer";
+import { ISharedObject, ISharedObjectEvents } from "./types";
+
+export interface ISharedObjectStateHost {
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObjectBase.ts:29]: I'm not sure what "host" refers to. I think ISharedObjectSnapshot would be appropriate name, but "snapshot" is so overloaded term that I do not want to use it here. Also, I think long term we want to extend that beyond DDS / SharedObject (base class). Looking for some name / synonyms that captures intent - IObjectRevision or IRevision ?

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 8]
```diff
@@ -3,450 +3,46 @@
...
-        this.snapshotCore(serializer);
-
-        // The GC data for this shared object contains a single GC node. The outbound routes of this node are the
-        // routes of handles serialized during snapshot.
-        return {
-            gcNodes: { "/": serializer.getSerializedRoutes() },
-        };
+    protected captureStateCore(): ISharedObjectStateHost {
+        const snapshot = this.snapshotCore(this.serializer);
+        return { summarize: async (serializer: IFluidSerializer, fullTree: boolean = false) => {
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObject.ts:36]: I'm a bit lost here RE use of serializer and intentions.
Looking at existing code (which is a bit hard to follow), I believe
- this.serializer was used outside of summaries only. And old code specifically was setting this._isSummarizing = true to ensure that any usage of it in summary would assert.
- summary code was always creating serializer on the fly and was passing it down, but otherwise would destroy it after summary.

I'd think this code needs to do the same - ensure that this.snapshotCore would be using passed in (newly allocated) serializer but would not be able to touch this.serializer. And thus, I'd think summarize() method should not have serializer argument?

I might be wrong, I'm simply trying to follow before & after code and see the differences, intentional or not.

File: packages/dds/shared-object-base/src/sharedObjectBase.ts [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+    public isAttached(): boolean {
+        return this.services !== undefined && this.runtime.attachState !== AttachState.Detached;
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).captureState}
+     */
+    public captureState(): IStateHost {
+        const state = this.captureStateCore();
+        return { summarize: async (fullTree: boolean = false) => {
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObjectBase.ts:204]: See my comment in SharedObject class RE usage of this._serializer.
But it feels like there is no need to wrap state returned from this.captureStateCore() - it should be exactly state returned to captureState() caller, and thus likely we can have captureState() abstract here (and no need for extra interfaces and Core function)

File: packages/dds/shared-object-base/src/sharedObjectBase.ts [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+     * other GC nodes.
+     * Derived classes must override this to provide custom list of references to other GC nodes.
+     */
+    protected async getGCDataCore(): Promise<IGarbageCollectionData> {
+        // We run the full summarize logic to get the list of outbound routes from this object. This is a little
+        // expensive but it's okay for now. It will be updated to not use full summarize and make it more efficient.
+        // See: https://github.com/microsoft/FluidFramework/issues/4547
+        const serializer = new SummarySerializer(this.runtime.channelsRoutingContext);
+        const state = this.captureStateCore();
+        await state.summarize(serializer, false);
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/sharedObjectBase.ts:242]: By the time SharedObject captured state, it's too late to give it a serializer. I think the right model is the following:
- captureStateCore (or captureState, not sure there is need for both here) would internally create a serializer and capture snapshot with it, keeping serializer
- The object returned has summarize() method that does not take serializer, and simply returns a summary.
- There is also another method, called getGCData(), that would use internally stored serializer to spit it out.

That way serializer is left to be implementation detail, and API is relatively high level:
- get a revision (aka state)
- ask revision to generate summary.
- ask revision to generate GC info

File: packages/dds/shared-object-base/src/types.ts [hunk 11]
```diff
@@ -31,10 +31,10 @@ export interface ISharedObject<TEvent extends ISharedObjectEvents = ISharedObjec
     isAttached(): boolean;
 
     /**
-     * Generates summary of the shared object.
-     * @returns A tree representing the summary of the shared object.
+     * Capture the current content of the channel to be used to generate a summary
+     * @returns Object containing captured state that exposes functionality over that state (e.g. produce summary)
      */
-    summarize(fullTree?: boolean, trackState?: boolean): ISummaryTreeWithStats;
+    captureState(): ISnaphost;
```
Comment (vladsud) [inline packages/dds/shared-object-base/src/types.ts:37]: Maybe I'm blind, but I can't figure out what is this "ISnaphost"
I see only one file changed in that package (packages/runtime/datastore-definitions/src/channel.ts) and it's not there (and not in repo).
Also is it a typo (i.e. should be ISnapshot)?

File: packages/dds/shared-object-base/src/sharedObject.ts [hunk 12]
```diff
@@ -450,3 +364,89 @@ export abstract class SharedObject<TEvent extends ISharedObjectEvents = ISharedO
...
+    {
+        super(id, runtime, attributes);
+
+        this._serializer = new FluidSerializer(this.runtime.channelsRoutingContext);
+    }
+
+    /**
+     * {@inheritDoc (ISharedObject:interface).captureRevision}
+     */
+    public captureRevision(): IChannelRevision {
```
Comment (agarwal-navin) [inline packages/dds/shared-object-base/src/sharedObject.ts:411]: I see a couple of problems with having this API instead of separate APIs for summarize and getGCData:
1. How will a DDS have custom implementation for one of these APIs? For example, SharedMatrix and SharedString work on a different set of data for generating GC state vs summary state. (Btw, this change will break them).
2. The state returned by summarize is affected by GC. When GC runs on the data returned by getGCData, it may update the "reference" state of certain nodes in the system. This information needs to be encoded in the summary. Additionally, some nodes which did not need summarization before (because its data did not change) may need to be summarized because its "reference" state changed.
This is not really a problem at this layer yet but as soon as we have sub-DDS GC, this will break. Also, IFluidDataStoreChannel's shape will also become similar to this and it will run into this problem now.

Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:411]: The main challenge with point 2 above is that we don't fully know whether we need to re-summarize a node or not until GC has run.

Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObject.ts:411]: 1. I'm not sure if my changes for sequence and matrix are the right thing. For DDSs that want to do something custom, they should derive from SharedObjectCore and fully implement captureRevision themselves. They'd still need to capture all the state necessary for both summaries and GC.
2. Is it possible to update the captured state and generate another summary off that? or does it need to happen on the sharedObject?

File: packages/runtime/container-runtime/src/dataStores.ts [hunk 13]
```diff
@@ -367,16 +367,17 @@ export class DataStores implements IDisposable {
         return summaryBuilder.getSummaryTree();
     }
 
-    public createSummary(): ISummaryTreeWithStats {
+    public async createSummary(): Promise<ISummaryTreeWithStats> {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStores.ts:370]: Same comment as above: It's possible to ensure (by code inspection) that this function returns state of container at the time when createSummary() call started. But we will not be able to sustain that over time, even with UTs. I believe if we go that route, we have to commit (and implement quickly as a follow up) same flow you are implementing for DDS here, i.e., for whole container, with synchronous "capture revision" and async "generate snapshot out of it.

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 14]
```diff
@@ -829,13 +829,13 @@ export class LocalFluidDataStoreContextBase extends FluidDataStoreContext {
         });
     }
 
-    public generateAttachMessage(): IAttachMessage {
+    public async generateAttachMessage(): Promise<IAttachMessage> {
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:832]: This API existed for two reasons:
1. synchronous nature of workflows that depended on it
2. some subtle differences in content returned from async summarize methods.
Given that # 1 is gone in this PR, it would be great to have a follow up issue to examine # 2 and see if we can substantially reduce code duplication. I.e. make this logic rely as much as possible on summarization flow, only adding (if needed) subtle differences that are required (if they are) for attachment workflows

File: packages/loader/container-loader/src/container.ts [hunk 15]
```diff
@@ -778,7 +778,7 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
             if (!hasAttachmentBlobs) {
                 // Get the document state post attach - possibly can just call attach but we need to change the
                 // semantics around what the attach means as far as async code goes.
-                const appSummary: ISummaryTree = this.context.createSummary();
+                const appSummary: ISummaryTree = await this.context.createSummary();
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:781]: Here is an example of the problem that I worry about when making createSummary() flow async. All capturing of state (aka revision) has to happen synchronously as otherwise we may not capture consistent state (i.e., parts of state would come from different points in time).
Here, we await container runtime summary creation, but capture protocol state and switch container to attaching state after that. That means that protocol state may change while we are blocked on await. Similar, any changes to container while we await would not generate ops because container is not switched to attaching state early enough, thus those changes would not make it to either snapshot nor ops that follow.

The only right way here is to ensure that all state is captured before any async activity is awaited. Ideally through 2-step process - get revision synchronously, generate snapshot from revision async. As a temp solution, we can do code inspection + UTs to get something working quickly (and fixing any cases like this one by ensuring that promise is captured early but await is pushed to line 793).  But this can't be long-term solution

File: packages/loader/container-loader/src/container.ts [hunk 16]
```diff
@@ -829,7 +829,7 @@ export class Container extends EventEmitterWithErrorHandling<IContainerEvents> i
                 }
 
                 // take summary and upload
-                const appSummary: ISummaryTree = this.context.createSummary(redirectTable);
+                const appSummary: ISummaryTree = await this.context.createSummary(redirectTable);
```
Comment (vladsud) [inline packages/loader/container-loader/src/container.ts:832]: Same problem here - await needs to be after this.emit("attached"). I'm not even sure code can handle it correctly today (i.e. async inflight summary process while we emit attached event) - maybe, but I'd need to do deeper code inspection to be sure, as we never have had such state.

File: packages/runtime/container-runtime/src/containerRuntime.ts [hunk 17]
```diff
@@ -1736,12 +1736,12 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
...
      * new storage IDs so requests can be redirected.
      */
-    public createSummary(blobRedirectTable?: Map<string, string>): ISummaryTree {
+    public async createSummary(blobRedirectTable?: Map<string, string>): Promise<ISummaryTree> {
         if (blobRedirectTable) {
             this.blobManager.setRedirectTable(blobRedirectTable);
         }
 
-        const summarizeResult = this.dataStores.createSummary();
+        const summarizeResult = await this.dataStores.createSummary();
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/containerRuntime.ts:1744]: Similar problem here (requiring changing existing flows / APIs) - addContainerBlobsToSummary() below should happen before await here, as we need to do all "initiation" of data collection synchronously in one go, to make sure it represents state at the same point in time.

File: packages/runtime/container-runtime/src/dataStoreContext.ts [hunk 18]
```diff
@@ -228,11 +228,11 @@ export abstract class FluidDataStoreContext extends TypedEventEmitter<IFluidData
         this._attachState = this.containerRuntime.attachState !== AttachState.Detached && this.existing ?
             this.containerRuntime.attachState : AttachState.Detached;
 
-        this.bindToContext = () => {
+        this.bindToContext = async () => {
             assert(this.bindState === BindState.NotBound, 0x13b /* "datastore context is already in bound state" */);
             this.bindState = BindState.Binding;
             assert(this.channel !== undefined, 0x13c /* "undefined channel on datastore context" */);
-            bindChannel(this.channel);
+            await bindChannel(this.channel);
```
Comment (vladsud) [inline packages/runtime/container-runtime/src/dataStoreContext.ts:235]: Similar to other places, it's extremely hard to say if conversion to async here would not open pandora box of bugs.
Given that this code is used in synchronous workflows (i.e., storing a handle as a value in a map - that will cause attachment of data store / DDS that are backed by such handle, which previously was fully synchronous process), it's hard to believe that async parts of this process can happen later, when user has a chance to continue to modify objects that we are trying to snapshot. Similar as in other places, we need to ensure we capture enough state (revisions) synchronously to represent state of universe before allowing async processes run off that revision.

File: packages/runtime/datastore/src/channelContext.ts [hunk 19]
```diff
@@ -69,12 +69,14 @@ export function createServiceEndpoints(
...
+// TODO: this needs to be split up into two calls so callers do all the captures first before summarizeState
+export async function summarizeChannel(
     channel: IChannel,
     fullTree: boolean = false,
     trackState: boolean = false,
-): ISummaryTreeWithStats {
-    const summarizeResult = channel.summarize(fullTree, trackState);
+): Promise<ISummaryTreeWithStats> {
+    const state = channel.captureSummaryState(fullTree);
+    const summarizeResult = await channel.summarizeState(state);
```
Comment (vladsud) [inline packages/runtime/datastore/src/channelContext.ts:79]: I'm a bit at a loss here.
You have added IChannel.captureRevision().summarize() flow.
Why is there some other captureSummaryState() flow? I do not see IChannel having captureSummaryState()
Is this simply non-updated code from prior revisions?

**Comment by vladsud** [comment]
Looking deeper at the set of changes, I think you will not be able to escape from making 2-step process (capture revision, generate summary) across all layers.
If you really want to avoid it in single PR, I think you will have to leave dual workflows (synchronous, and async) across all layers, and simply introduce async flow (the way you are doing) to DDS (while leaving sync flow completely intact).
It just too easy to make mistakes and capture state from different points in time, and it will take substantial effort to write all the required UTs to get confidence that such massive change across all layers actually does not regress anything.

File: packages/runtime/container-runtime/src/dataStoreContext.ts (hunk 18 above)
Comment (scarlettjlee) [reply packages/runtime/container-runtime/src/dataStoreContext.ts:235]: I don't understand the full implications of this. Does it mean that it will be very hard to get things right? Or we should enforce that some things remain synchronous?

File: packages/runtime/datastore/src/channelContext.ts (hunk 19 above)
Comment (scarlettjlee) [reply packages/runtime/datastore/src/channelContext.ts:79]: Yes, non-updated. The last couple commits were just to get feedback on the change to SharedObject before changing it everywhere.

File: packages/dds/shared-object-base/src/sharedObjectBase.ts (hunk 10 above)
Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObjectBase.ts]: As @agarwal-navin [comments](https://github.com/microsoft/FluidFramework/pull/8422#discussion_r770752289), running GC can change what should be in the summary. Given that the potential change is only to say whether something is referenced or not, it seems possible to just fix that up in the IChannelRevision. That feels like a strange thing to do though.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 8 above)
Comment (scarlettjlee) [reply packages/dds/shared-object-base/src/sharedObject.ts]: The changes are deliberate. We only need to ensure that only the particular, new SummarySerializer is used during GC and never any other serializer. It doesn't matter during non-GC summarization. I verified with Navin.

**Comment by vladsud** [comment]
A thought worth sharing RE createSummary() flow: I think there is an incremental path how to convert everything to single flow.
What if we leave this PR for now, and start another one (that would be pre-game for this one). It would change createSummary() to return an IChannelRevision instead of current summary. Under the covers, it can capture snapshot IChannelRevision would have just one function (possibly even synchronous to make changes incremental) to expose such snapshot. That way we can rather easily (only one complication - back compat) changes all existing layers to work with IChannelRevision with no other semantic or behavior changes. Then we can make further steps like
- async summary on IChannelRevision - will require careful code inspection to make sure all awaits are after ALL state has been captured (some of the comments in this PR)
- introduction of SharedObjectCore and allowing truly async behavior where it matters.

RE back compat: it's hard as we require back compat across two boundaries - Loader (Container + ContainerContext) to IRuntime (ContainerRuntime) and container runtime (IContainerRuntimeBase) to data store (IFluidDataStoreChannel). We will need parallel system (i.e. support old functions with old semantics while having new flow). Code can likely be reused but shape of API will need to support both for a while. Thus we would need createSummary2 in some format (i.e. can't just reuse existing name, unless there is something else telling each layer how to think about shape of this API)

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 12 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:411]: I was asking myself that question (that Navin raised) - can summary and gc data come from same state or not, sounds not (sorry I forgot it). I think there are two ways to address it:
- Leave API as is but capture revision twice (at different times).
- Leave GC API the same, while revision to support only summary API.

It's possible that the answer will be different on the layer. I.e. the general structure should likely allow flexibility, and this we might want to leave GC API as is (including on SharedObjectCore layer). But if all but Whiteboard DDSs can always generage GC from summary, then I'd argue we want to build code in a way where this duplication of logic happens in single place (SharedObject) by having implementation of IChannelRevision here that can extract GC data out of it (i.e. have concrete implementation of IChannelRevision that captures and exposes GC data). That way, SharedObject.getGCData() can just do this.captureRevision().getGcData().

**Comment by agarwal-navin** [comment]
I have a question about the first step, i.e., synchronously getting summary state of a node, in the proposed 2-step summarization process:
How will this work at layers such as data store context and channel context with delay loading? For example, if a data store has not yet been realized during summary and it has changed, we need to realize it first and so we cannot capture its summary state synchronously.

**Comment by agarwal-navin** [comment]
This will work for the summary generated for detached container via `createSummary` but not for the regular summarization flow.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:411]: I don't think we can implement it via the first way you mentioned. The reason is that the summary state of a node may have the reference state (whether its unreferenced or not) based on an older summary state. Basically, GC state can be from seq#100 whereas summary state can be from seq#200. We cannot mark a node at seq#200 referenced / unreferenced based on its data from seq#100 - that would be inconsistent.

Both the summary and GC state in a summary should be off the same point in time (seq#).

**Comment by vladsud** [comment]
@agarwal-navin, that's correct, today's captureSummary does not work for those cases. New flow will have exactly same limitations.
I think you are asking how we extend it to be the only summary flow in container, and make it work across all those different states that have different assumptions. I think the only answer - very slowly, and only when work in layers underneath layers is done to support it, and only if/when it happens. I.e. nothing prevents us from writing a bunch of code that will capture revision of current state for data store at any state. This is likely tremendous amount of work as we would need to duplicate current state / tree completely to isolate it from "live" state. Or bake immutability across layers. But when / if this is done, we can return revision in some form and work on it.

Note that we do not have to do it. We can continue to have two flows - async summary (as today) and sync capture revision, with revision flow working only where captureSummary() works today. It will still give us a boost in the form of leaf nodes (DDS) being able to switch to 2-step approach when generating summaries (i.e. DDSs will have only one flow to generate summary - 2-step process through revisions). This will solve Whiteboard problem, while maintaining correctness of the system. It's correctness that I'm after, plus at least one tiny place (DDSs) where we have only one way of doing summaries (not two as we have at data store layer), with possible long future capability of having one way of doing summaries across all layers.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 12 above)
Comment (vladsud) [reply packages/dds/shared-object-base/src/sharedObject.ts:411]: I'm not sure how to interpret your comment :) DDS just exposes ability to capture these states, at any moment in time. External process governs when these methods are called - for same state or not. DDS itself does not capture any additional state, so there is no differences here. The only thing that differs is how internally it captures GC - if all SharedObject derived classes extract GC data in exactly same way (by essentially summarizing themselves), then there is no need to duplicate this code - SharedObject base class can help here.

**Comment by agarwal-navin** [comment]
Okay, I think I misunderstood this change. This is only changing the synchronous captureSummary for detached container by making it async. It is not changing the regular summarization flow that happens in the summarizer client. And we are looking for ways to make the DDS implement a single way of generating the summary for both the above flows.

File: packages/dds/shared-object-base/src/sharedObject.ts (hunk 12 above)
Comment (agarwal-navin) [reply packages/dds/shared-object-base/src/sharedObject.ts:411]: I misunderstood this change. I was under the impression that we are also changing the way summarization happens in the regular case in summarizer client.

**Comment by vladsud** [comment]
So I feel like PRs need to flow in this order:
1. Introduce async summary flow to DDS, but leave sync flow as is
  - Do not touch any other layer
  - async flow for SharedObject is based on sync flow (i.e., all non-Whiteboard DDSs implement just one method). Whiteboard has to deal with two flows for now.
2. Introduce capturing revisions for all layers. For now, only works in detached state. Allows to remove 1b limitation, but keep two summary flows at higher levels
3. Rework layers to be able to capture revisions at any moment in time (attached or not). Allows to unify summarization process across all layers.

# 1 is rather simple and immediately unblocks Whiteboard (even though in an ugly / hard to deal / explain way)
# 2 is not that hard, but back-compat requirements make it more problematic
# 3 is likely a ton of work. We would need to estimate it and make decision if it's worth it.

Note that # 2 gives us tool that can be used in other flows (and can be improved over time). I.e. ability to capture revision at DDS level is very powerful for all kinds of things (would be useful today for Word's transactions request). That said # 3 is what make is really powerful :)

**Comment by agarwal-navin** [comment]
That sounds like a good plan. Breaking this down into stages will definitely make it simpler to identify and catch issues.

**Comment by scarlettjlee** [comment]
This may be reopened later. For now, it's replaced by #8592.
